instance/tasks.py

```diff
@@ -78,45 +76,71 @@ def start_deployment(
 
     if deployment.cancelled:
         logger.info('Deployment %s was cancelled, returning.', deployment.id)
-        return False
+        return
 
     logger.info('Spawning servers for deployment %s [%s]', deployment, deployment.id)
+    old_server_ids = list(instance.appserver_set.filter(_is_active=True).values_list('id', flat=True))
     # Launch configured number of appservers for instance
-    appserver_spawn_tasks = spawn_appserver.map(
-        (instance_ref_id, mark_active_on_success, False, num_attempts, success_tag, failure_tag, deployment.id)
-        for _ in range(instance.openedx_appserver_count)
-    )
+    for _ in range(instance.openedx_appserver_count):
+        spawn_appserver(
+            instance_ref_id=instance_ref_id,
+            mark_active_on_success=mark_active_on_success,
+            num_attempts=num_attempts,
+            success_tag=success_tag,
+            failure_tag=failure_tag,
+            deployment_id=deployment.id,
+            old_server_ids=old_server_ids,
+            target_count=instance.openedx_appserver_count,
+        )
 
-    appserver_ids = appserver_spawn_tasks.get(blocking=True)
 
-    if not all(appserver_ids):
-        return False
+@db_task()
+def check_deactivation(
+        result: bool,
+        instance_ref_id: int = None,
+        deployment_id: int = None,
+        old_server_ids: List[int] = None,
+        target_count: int = None,
+):
+    """
+    Called in pipeline with make_appserver_active when the current deployment is meant to deactivate old servers
+    upon success. Since any particular server could be the last one, this checks to make sure that the correct number
+    of servers which are new are now available, and that the current deployment hasn't been cancelled.
 
-    # For redundancy - if app server deployment was successful even after cancelling, we prevent the change in
-    # active appserver here
+    If this is the case, it deactivates all the old app servers.
+    """
+    if not result:
+        # This function is chained after mark_appserver_active. If that function failed, we already know things aren't
+        # OK and bail out.
+        return
+    # One last chance to prevent activation from switching over, in case some other deployment is now taking precedence.
     deployment = OpenEdXDeployment.objects.get(pk=deployment_id)
     if deployment.cancelled:
         logger.info('Deployment %s was cancelled, returning.', deployment.id)
-        return False
+        return
 
+    instance = OpenEdXInstance.objects.get(ref_set__pk=instance_ref_id)
     # If this deployment is to be marked active on success, others should be deactivated automatically
-    if mark_active_on_success:
-        other_appservers = instance.appserver_set.filter(_is_active=True).exclude(pk__in=appserver_ids)
+    current_count = instance.appserver_set.exclude(id__in=old_server_ids).filter(_is_active=True).count()
+    all_ready = current_count == target_count
+    if all_ready:
```

Comment:
@fox, it will be useful to log a message when this case isn't true, for example when 1 of the 2 expected appservers have been spawned,